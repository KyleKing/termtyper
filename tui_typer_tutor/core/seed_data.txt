def __init__(self, *args, **kwargs):<Return>super().__init__()<Return>self._counter = 0<Esc>
YY
ciw
vt"p
Vkky
<Space>ld
'
!
"
#
$
%
&
(
)
*
+
,
-
.
/
:
;
<Bar>
<Bslash>
<Return>
<Tab>
=
?
@
[
]
^
_
{
}
~
}
