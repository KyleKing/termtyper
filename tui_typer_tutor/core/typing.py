"""Typing Logic."""

import bisect

from beartype import beartype
from pydantic import BaseModel, Field

from ..constants import VIM_TO_TEXTUAL


class ExpectedKey(BaseModel):
    """Expected Key."""

    raw: str
    """Textual Key Name."""

    @property
    def text(self) -> str:
        """Displayed text."""
        return VIM_TO_TEXTUAL.inverse.get(self.raw) or self.raw

    @property
    def width(self) -> int:
        """Displayed text."""
        return len(self.text)


class TypedKey(ExpectedKey):
    """Typed Key."""

    expected: ExpectedKey
    """Store the expected key when typed and expected become out-of-sync."""

    @property
    def was_correct(self) -> bool:
        """If typed key matches expected."""
        return self.text == self.expected.text


class _KeysAccum(BaseModel):
    """Histogram-style store for identifying character widths."""

    expected: list[int] = Field(default_factory=list)
    """Accumulating count of character width per Key."""

    typed: list[int] = Field(default_factory=list)
    """Accumulating count of character width per Key."""

    @beartype
    def get_indices(self, accum: list[int], start: int, end: int) -> tuple[int, int]:
        """Return the adjusted start and end."""
        adj_start = bisect.bisect(accum, start)
        adj_end = bisect.bisect(accum, end - start + adj_start) if len(accum) > end else len(accum)
        return (adj_start, adj_end)


class Keys(BaseModel):
    """Key Model."""

    expected: list[ExpectedKey] = Field(default_factory=list)
    """The expected keys for practice."""

    typed_all: list[TypedKey] = Field(default_factory=list)
    """Append-only."""

    typed: list[TypedKey] = Field(default_factory=list)
    """Only tracks non-deleted keys."""

    last_was_delete: bool = False
    """Indicate if last operation was a delete."""

    accum: _KeysAccum = Field(default_factory=_KeysAccum)
    """Accumulating index store (private)."""

    def copy(self) -> 'Keys':
        """Ensure that the lists are deep-copied."""
        keys = super().copy()
        keys.expected = [*keys.expected]
        keys.typed_all = [*keys.typed_all]
        keys.typed = [*keys.typed]
        keys.accum.expected = [*keys.accum.expected]
        keys.accum.typed = [*keys.accum.typed]
        return keys

    @beartype
    def get_expected(self, start: int, end: int) -> list[ExpectedKey]:
        """Retrieve the next page of expected keys."""
        if not self.accum.expected:  # Lazily evaluate the expected values
            counter = 0
            for exp in self.expected:
                counter += exp.width
                self.accum.expected.append(counter)
        adj_start, adj_end = self.accum.get_indices(self.accum.expected, start, end)
        return self.expected[adj_start:adj_end]

    @beartype
    def get_typed(self, start: int, end: int) -> list[TypedKey]:
        """Retrieve the next page of typed keys."""
        adj_start, adj_end = self.accum.get_indices(self.accum.typed, start, end)
        return self.typed[adj_start:adj_end]

    @beartype
    def store(self, key: TypedKey) -> None:
        """Store a new typed key."""
        self.typed_all.append(key)
        self.last_was_delete = key.raw == 'backspace'
        if self.last_was_delete:
            if self.typed:
                self.typed = self.typed[:-1]
                self.accum.typed.pop()
        else:
            self.typed.append(key)
            counter = (self.accum.typed or [0])[-1]
            self.accum.typed.append(counter + key.width)


@beartype
def on_keypress(raw: str, keys: Keys) -> None:
    """Process a key press."""
    expected = keys.expected[len(keys.typed)]
    key = TypedKey(raw=raw, expected=expected)
    keys.store(key)
