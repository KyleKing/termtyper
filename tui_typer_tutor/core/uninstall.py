"""Uninstall files managed by `ttt`."""

import shutil
from pathlib import Path

import platformdirs
from beartype import beartype
from corallium.log import logger

from .. import APP_NAME


@beartype
def get_cache_dir() -> Path:
    """Application cache directory."""
    return Path(platformdirs.user_cache_dir(APP_NAME))


@beartype
def uninstall() -> None:
    """Uninstall files managed by `ttt`."""
    cache_dir = get_cache_dir()
    if cache_dir.is_dir():
        logger.info('Removing cache directory', cache_dir=cache_dir)
        shutil.rmtree(cache_dir)
