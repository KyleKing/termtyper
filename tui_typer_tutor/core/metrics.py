"""Manage Metrics."""

import csv
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo

from beartype import beartype
from pydantic import BaseModel

from .typing import Keys


@beartype
def utcnow() -> datetime:
    """Return the current time in UTC."""
    return datetime.now(tz=ZoneInfo('UTC'))


class SessionMetrics(BaseModel):
    """Session metrics."""

    filename: str
    session_start: datetime
    session_end: datetime | None = None
    typed_correct: int = 0
    typed_incorrect: int = 0

    @classmethod
    def from_filename(cls, filename: str) -> 'SessionMetrics':  # noqa: RBT002
        """Initialize Metrics based on the filename."""
        return cls(filename=filename, session_start=utcnow())

    def end_session(self, keys: Keys) -> 'SessionMetrics':  # noqa: RBT002
        """Update the typed counters based on `Keys`."""
        self.session_end = utcnow()
        for key in keys.typed_all:
            if key.expected:
                if key.was_correct:
                    self.typed_correct += 1
                else:
                    self.typed_incorrect += 1
        return self


# PLANNED: Support Linux/Windows
CSV_PATH = Path.home() / '.config/tui-typer-tutor/metrics.csv'


@beartype
def append_csv(metrics: SessionMetrics) -> None:
    """Write metrics to the global CSV."""
    csv_columns = ['filename', 'session_start', 'session_end', 'typed_correct', 'typed_incorrect']
    if not CSV_PATH.is_file():
        CSV_PATH.parent.mkdir(exist_ok=True, parents=True)
        with CSV_PATH.open(mode='w', newline='', encoding='utf-8') as _f:
            csv.writer(_f).writerow(csv_columns)  # nosemgrep

    ser_metrics = metrics.dict()
    metrics_row = [ser_metrics[_c] for _c in csv_columns]
    with CSV_PATH.open('a', newline='', encoding='utf-8') as _f:
        csv.writer(_f).writerow(metrics_row)  # nosemgrep
