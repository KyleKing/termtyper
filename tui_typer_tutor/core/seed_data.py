"""Load Seed Data."""

import random

from beartype import beartype

from ..constants import DISPLAY_TO_TEXTUAL
from .typing import ExpectedKey


@beartype
def load_seed_data(seed_text: str) -> list[ExpectedKey]:
    """Load Seed Data in Vim format."""
    grouped_keys = [
        [ExpectedKey(textual=DISPLAY_TO_TEXTUAL[token]) for token in line]
        for line in seed_text.split('\n')
    ]
    random.shuffle(grouped_keys)  # noqa: DUO102
    return [_k for _keys in grouped_keys for _k in _keys]
