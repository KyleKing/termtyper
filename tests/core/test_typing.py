import pytest

from tui_typer_tutor.core.typing import (
    ExpectedKey,
    Keys,
    on_keypress,
)

from ..seed_data import lol_keys


@pytest.mark.parametrize(
    ('raws', 'keys'),
    [
        (['ctrl+q'], Keys(expected=[ExpectedKey(textual='s')])),
        (['t'], lol_keys()),
        (['backspace'], lol_keys()),
        (['backspace', 'backspace', 'l', 'o', 'l'], lol_keys()),
    ],
)
def test_on_keypress(
    raws: str,
    keys: Keys,
    snapshot,
):
    for raw in raws:
        on_keypress(raw, keys)  # act

    assert keys == snapshot


@pytest.mark.parametrize(
    ('raw', 'was_correct'),
    [
        ('0', False),
        ('o', True),
    ],
)
def test_on_keypress___extended(
    *,
    raw: str,
    was_correct: bool,
    snapshot,
):
    keys = lol_keys()

    on_keypress(raw, keys)  # act

    assert keys.typed[-1].was_correct == was_correct
    assert keys == snapshot
