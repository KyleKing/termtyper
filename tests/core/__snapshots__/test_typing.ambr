# serializer version: 1
# name: test_on_keypress[raws0-keys0]
  Keys(expected=[ExpectedKey(textual='s')], typed_all=[TypedKey(textual='ctrl+q', expected=ExpectedKey(textual='s'))], typed=[TypedKey(textual='ctrl+q', expected=ExpectedKey(textual='s'))], last_was_delete=False)
# ---
# name: test_on_keypress[raws1-keys1]
  Keys(expected=[ExpectedKey(textual='l'), ExpectedKey(textual='o'), ExpectedKey(textual='l')], typed_all=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='t', expected=ExpectedKey(textual='o'))], typed=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='t', expected=ExpectedKey(textual='o'))], last_was_delete=False)
# ---
# name: test_on_keypress[raws2-keys2]
  Keys(expected=[ExpectedKey(textual='l'), ExpectedKey(textual='o'), ExpectedKey(textual='l')], typed_all=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='backspace', expected=None)], typed=[], last_was_delete=True)
# ---
# name: test_on_keypress[raws3-keys3]
  Keys(expected=[ExpectedKey(textual='l'), ExpectedKey(textual='o'), ExpectedKey(textual='l')], typed_all=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='backspace', expected=None), TypedKey(textual='backspace', expected=None), TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='o', expected=ExpectedKey(textual='o')), TypedKey(textual='l', expected=ExpectedKey(textual='l'))], typed=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='o', expected=ExpectedKey(textual='o')), TypedKey(textual='l', expected=ExpectedKey(textual='l'))], last_was_delete=False)
# ---
# name: test_on_keypress___extended[0-False]
  Keys(expected=[ExpectedKey(textual='l'), ExpectedKey(textual='o'), ExpectedKey(textual='l')], typed_all=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='0', expected=ExpectedKey(textual='o'))], typed=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='0', expected=ExpectedKey(textual='o'))], last_was_delete=False)
# ---
# name: test_on_keypress___extended[o-True]
  Keys(expected=[ExpectedKey(textual='l'), ExpectedKey(textual='o'), ExpectedKey(textual='l')], typed_all=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='o', expected=ExpectedKey(textual='o'))], typed=[TypedKey(textual='l', expected=ExpectedKey(textual='l')), TypedKey(textual='o', expected=ExpectedKey(textual='o'))], last_was_delete=False)
# ---
